ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,907 @@
+		if validator_accounts.is_empty() {
+			self.broadcast_private_transaction(rlp.into());
+			return Ok(());
+		}
+		let private_tx: PrivateTransaction = UntrustedRlp::new(rlp).as_val()?;
+		let contract = private_tx.contract;
+		let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+
+		match contract_validators
+			.iter()
+			.filter(|&&address| validator_accounts
+				.iter()
+				.any(|&validator| validator == address))
+			.next() {
+			None => {
+				// Not for verification, broadcast further to peers
+				self.broadcast_private_transaction(rlp.into());
+				return Ok(());
+			},
+			Some(&validation_account) => {
+				trace!("Private transaction taken for verification");
+				let client = self.client.read().clone().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+				let client = client.upgrade().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+				let original_tx = self.extract_original_transaction(private_tx.clone(), &contract)?;
+				trace!("Validating transaction: {:?}", original_tx);
+				let details_provider = TransactionDetailsProvider::new(&*client as &MiningBlockChainClient);
+				let insertion_time = client.chain_info().best_block_number;
+				// Verify with the first account available
+				trace!("The following account will be used for verification: {:?}", validation_account);
+				let add_res = self.transactions_for_verification.lock().add_transaction(original_tx, contract, validation_account, private_tx.hash(), &details_provider, insertion_time);
+				match add_res {
+					Ok(_) => {
+						let channel = client.get_io_channel();
+						let channel = channel.lock();
+						channel.send(ClientIoMessage::NewPrivateTransaction)
+							.map_err(|_| PrivateTransactionError::ClientIsMalformed.into())
+					},
+					Err(err) => Err(err),
+				}
+			}
+		}
+	}
+
+	/// Private transaction for validation added into queue
+	pub fn on_private_transaction_queued(&self) -> Result<(), EthcoreError> {
+		self.process_queue()
+	}
+
+	/// Retrieve and verify the first available private transaction for every sender
+	fn process_queue(&self) -> Result<(), EthcoreError> {
+		let client = self.client.read().clone().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+		let client = client.upgrade().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+		let ready_transactions = self.transactions_for_verification.lock().ready_transactions();
+		let fetch_nonce = |a: &Address| client.latest_nonce(a);
+		for transaction in ready_transactions {
+			let transaction_hash = transaction.hash();
+			match self.transactions_for_verification.lock().private_transaction_descriptor(&transaction_hash) {
+				Ok(desc) => {
+					match self.config.read().validator_accounts.iter().find(|&&account| account == desc.validator_account) {
+						Some(account) => {
+							let accounts = self.accounts.read().clone().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
```

Comment:
That's pretty deeply nested. Perhaps it's possible to use some early exists instead?